### PR TITLE
chore(stages): Helper for walking over table range

### LINKED ISSUE #702: Clean up DB methods in stages
Go through the DB interactions in stages and clean up repeating/redundant methods where possible by adding methods to `Transaction`.

Some examples of code to clean up:
https://github.com/paradigmxyz/reth/blob/93b6ce597ad192e6ee6b7ef60fdaa030d0a16a51/crates/stages/src/stages/total_difficulty.rs#L67-L69
https://github.com/paradigmxyz/reth/blob/eb600d1f029da7381bd7f5f69a75aeafb9a54247/crates/stages/src/stages/execution.rs#L107-L111

### LINKED ISSUE #912: Helper for walking over table range
Currently, in order to walk a range we use the `cursor` -> `walk` -> take_while` pattern. This is nice, but repetiitve and footgun-y. e.g. https://github.com/paradigmxyz/reth/blob/5d45325e43afaf1e7cd60142ae764494db105602/crates/stages/src/stages/execution.rs#L103-L107

```
tx.cursor_read::<tables::StorageChangeSet>()?
            .walk((from_transition_rev, H160::zero()).into())?
            .take_while(|res| {
                res.as_ref()
                    .map(|(TransitionIdAddress((k, _)), _)| *k < to_transition_rev)
                    .unwrap_or_default()
            })
```

It would be nice to instead have a helper which gives us a walker over the range like: `tx.walker_read::<T: Table>(start: impl Into<T::Key>, end: impl Into<T::Key>)` which abstracts that way.

### PR DESCRIPTION
**Pull request by jinsankim**
Closes #912 and closes #702

## Motivation
* Revise and clean up repetitive pattern, `cursor -> walk -> take_while`

### CODE REVIEW

File: crates/stages/src/db.rs [hunk 1]
```diff
@@ -164,6 +164,26 @@ where
         Ok((prev_body.start_tx_id + prev_body.tx_count, last_transition))
     }
 
+    /// Get the result vector with range [start_key:end_key).
+    /// Please note that start_key is inclusive and end_key is expclusive.
+    #[allow(clippy::type_complexity)]
+    pub(crate) fn get_range<T: Table>(
```
Comment (jinsankim) [inline crates/stages/src/db.rs:170]: I'm think about the name of this function. First, I started it with `walk_range()`. But I changed it as `get_range()` because it returns `Vector` but doesn't return `Iterator`. HDYT?

File: crates/stages/src/db.rs [hunk 2]
```diff
@@ -164,6 +164,26 @@ where
         Ok((prev_body.start_tx_id + prev_body.tx_count, last_transition))
     }
 
+    /// Get the result vector with range [start_key:end_key).
+    /// Please note that start_key is inclusive and end_key is expclusive.
+    #[allow(clippy::type_complexity)]
```
Comment (jinsankim) [inline crates/stages/src/db.rs:169]: I've got the clippy warning, `very complex type used. Consider factoring parts into type definitions` for the return type, `Result<Vec<(T::Key, T::Value)>, Error>`.

If defining the type, `type VecPairResult<T> = Result<Vec<(<T as Table>::Key, <T as Table>::Value)>, Error>;`, I've get the error, `error[E0658]: inherent associated types are unstable`.

How to resolve it? Please could anyone give me an advice?

**Comment by jinsankim** [comment]
retest

File: crates/stages/src/db.rs [hunk 3]
```diff
@@ -164,6 +165,21 @@ where
         Ok((prev_body.start_tx_id + prev_body.tx_count, last_transition))
     }
 
+    /// Get the result vector with range [start_key:end_key).
+    /// Please note that start_key is inclusive and end_key is expclusive.
```
Comment (onbjerg) [inline crates/stages/src/db.rs:169]:
```suggestion
    /// Get the result vector with range `[start_key:end_key)`.
    /// Please note that `start_key` is inclusive and `end_key` is exclusive.
```

File: crates/stages/src/stages/hashing_storage.rs [hunk 4]
```diff
@@ -91,62 +91,60 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
...
-                .fold(
-                    BTreeMap::new(),
-                    |mut accounts: BTreeMap<Address, BTreeMap<H256, U256>>,
-                     (TransitionIdAddress((_, address)), storage_entry)| {
-                        accounts
-                            .entry(address)
-                            .or_default()
-                            .insert(storage_entry.key, storage_entry.value);
-                        accounts
+            tx.get_range::<tables::StorageChangeSet>(
```
Comment (onbjerg) [inline crates/stages/src/stages/hashing_storage.rs:94]: I would decouple this into smaller code bites to make it more readable

File: crates/stages/src/stages/hashing_storage.rs [hunk 5]
```diff
@@ -165,41 +163,37 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
...
-            .try_for_each(|((address, key), value)| -> Result<(), StageError> {
-                hashed_storage.seek_by_key_subkey(address, key)?;
-                hashed_storage.delete_current()?;
-
-                if value != U256::ZERO {
-                    hashed_storage.append_dup(address, StorageEntry { key, value })?;
-                }
-                Ok(())
-            })?;
+        tx.get_range::<tables::StorageChangeSet>(
```
Comment (onbjerg) [inline crates/stages/src/stages/hashing_storage.rs:166]: Same here

**Review by onbjerg — CHANGES_REQUESTED** [review]
Looks good overall! Some nits on the usage of the API

**Comment by jinsankim** [comment]
resolved merge conflicts.

File: crates/stages/src/stages/hashing_storage.rs (hunk 4 above)
Comment (jinsankim) [reply crates/stages/src/stages/hashing_storage.rs]: I tried to resolve it w/ 89e761819bf4061735a80b13ea26048c4d265d9d. I just decouple the starting part because I'm not familiar with whole function chains. Is it acceptable?

File: crates/stages/src/db.rs [hunk 6]
```diff
@@ -164,6 +165,21 @@ where
...
+    where
+        T: Table,
+    {
+        self.cursor_read::<T>()?
+            .walk(start_key)?
+            .take_while(|res| match res {
+                Ok((key, _)) => *key < end_key,
+                Err(_) => false,
+            })
+            .collect::<Result<Vec<_>, _>>()
```
Comment (rkrasiuk) [inline crates/stages/src/db.rs:180]: perf: this would load everything into mem as opposed to walking and reading one by one, right @rakita? maybe we should return `Vec<PairResult<_>>`

Comment (jinsankim) [reply crates/stages/src/db.rs:180]: Yes, to my understanding, it loads everything into mem. As far as I know, if we do `collect()`, it might load everything into mem even though we return `Vec<PairResult<_>>`.

For perf, I tried to return `Walker<>` but failed because it's part of `DbCursor<>`.

### Positive thinking for my changes even though considering perf,
* All use cases that already do `collect()` except `sender_recovery.rs` that does `chunks()` iterator.
* We could remove this change from `sender_recovery.rs`.

### Negative thinking for my changes considering perf,
* I think fundamental solution is that revises `DbCursor<>` to add an interface
```
    fn walk_while<'cursor>(
        &'cursor mut self,
        start_key: T::Key,
        end_key: T::Key,
    ) -> Result<Walker<'cursor, 'tx, T, Self>, Error>
```    
* This approach is needed to change `DbCursor<>` and it could broadly affect on reth. 
* Just as a newbie contributor, I'm little bit affraid to approach this method.

I need feedback and opinion. Thanks,

Comment (jinsankim) [reply crates/stages/src/db.rs:180]: And also please note my comment on the function name, https://github.com/paradigmxyz/reth/pull/942#discussion_r1082614212

Comment (onbjerg) [reply crates/stages/src/db.rs:180]: @rkrasiuk returning `Vec<PairResult<_>>` would still result in this loading everything into mem.

I think the solution is to move this to the cursor as `walk_range` instead. It would require us to do `tx.cursor_read::<Table>().walk_range(..)` as opposed to `tx.get_range` but imo this is ok given perf concern.

File: crates/stages/src/stages/hashing_storage.rs (hunk 4 above)
Comment (onbjerg) [reply crates/stages/src/stages/hashing_storage.rs]: Ah, I see, the long chain existed prior to this PR too, I thought it was added because the diff lit all of it up. No worries! We should refactor it in a different PR though, it is hard to parse :smile:

File: crates/stages/src/db.rs (hunk 6 above)
Comment (rakita) [reply crates/stages/src/db.rs:180]: I am using it like that, I know how much I want to load and when reading the table I read it in batches to store it in memory.

Comment (onbjerg) [reply crates/stages/src/db.rs:180]: How come? You could just use the iterator directly in the storage hashing stage which would be way easier on memory compared to loading everything into a `Vec` that is immediately turned into an iterator again

Comment (onbjerg) [reply crates/stages/src/db.rs:180]: For reference I am talking about these lines: https://github.com/paradigmxyz/reth/blob/a1ad688c52fe6b61b07be36628589fbd96c6c1db/crates/stages/src/stages/hashing_storage.rs#L94-L114

The collect on L99 can be avoided, the one on L114 can't (but the memory footprint there is potentially *a lot* smaller)

Comment (onbjerg) [reply crates/stages/src/db.rs:180]: Similarly the `collect` can be avoided here: https://github.com/paradigmxyz/reth/blob/a1ad688c52fe6b61b07be36628589fbd96c6c1db/crates/stages/src/stages/hashing_storage.rs#L118-L130

Comment (rakita) [reply crates/stages/src/db.rs:180]: It is faster to load in bundles and have an iterator load as much as possible.

**Comment by jinsankim** [comment]
rebase to resolve merge conflicts.

**Comment by jinsankim** [comment]
This PR is replaced by #1025
